Make sure everything in the Cupertino page transition can be linear when back swiping

xster opened the pull request:
## Description

Currently, the route transition can become disconnected between the top and the bottom pages of a Cupertino transition because they can follow different curves. They should both be linearToEaseOut if animated but both be linear if driven by a back swipe. 

We should be checking for the participating animations for whether they're part of a swipe rather than checking the route.

## Related Issues

#28529

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes tests for *all* changed/updated/fixed behaviors (See [Test Coverage]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]).
- [x] No, this is *not* a breaking change.

xster committed:
Make sure everything in the Cupertino page transition can be linear when back swiping

xster commented:
@HansMuller want to get your feedback before I add the tests

xster committed:
lint

HansMuller commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -262,13 +271,28 @@ class CupertinoPageRoute<T> extends PageRoute<T> {
+  static Animation<double> _animationControllerForTransitionAnimation(Animation<double> animation) {
+    Animation<double> controller = animation;
+    while(controller is ProxyAnimation || controller is TrainHoppingAnimation) {
+      final Animation<double> proxy = controller;
```
This works because we believe that the implementation only creates these two animation types? If someone changed the implementation and started using FooAnimationWrapper or, worse still _FooAnimationWrapper, we'd just silently fail, right?

xster replied:
Took a different approach

HansMuller commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -188,12 +188,21 @@ class CupertinoPageRoute<T> extends PageRoute<T> {
-  static final Set<PageRoute<dynamic>> _popGestureInProgress = Set<PageRoute<dynamic>>();
+  static bool isPopGestureInProgress(PageRoute<dynamic> route) {
+    return _popGestureInProgress.contains(_animationControllerForTransitionAnimation(route.animation))
+        || _popGestureInProgress.contains(_animationControllerForTransitionAnimation(route.secondaryAnimation));
```
I'm not sure why _animationControllerForTransitionAnimation will always return route.secondaryAnimation if a route is using it. It seems like it would in simple cases, but all cases?

xster replied:
Took a different approach

xster committed:
add some tests

xster committed:
test

xster committed:
Change mechanism to use navigator.userGestureInProgress instead

xster commented:
So I took a different approach and didn't need to add a previous route tracker on the route itself. 

Navigator has a userGestureInProgress that I can use. Removed a static from the routes class. Added a bunch of tests. Also fixed https://github.com/flutter/flutter/pull/28756/files#r262295964. It diffbases on top of https://github.com/flutter/flutter/pull/28855.

HansMuller commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -188,12 +188,21 @@ class CupertinoPageRoute<T> extends PageRoute<T> {
 
   /// True if a Cupertino pop gesture is currently underway for [route].
   ///
+  /// This returns true for both the top route and the bottom route of a
```
This is still confusing since exactly what "pop gesture" means isn't self evident. Since we're just returning the value of `route.navigator.userGestureInProgress` you could just say as much and provide an explanation there.

xster replied:
done

HansMuller commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -646,9 +658,9 @@ class _CupertinoBackGestureController<T> {
       controller.removeStatusListener(_handleStatusChanged);
     }
     _animating = false;
+    onEnded();
```
Not sure why we now want to do this first. Previously it seemed like effectively an else clause. Why would we want to unconditionally dispose the route before removing it?

xster replied:
I thought I wrote it somewhere in the PR but didn't. This is a bug on everything downstream from didStopUserGesture such as NavigatorObserver.didStopUserGesture or NavigatorState.userGestureInProgress. It never worked because the route would be disposed before the back controller would be disposed. We were just lucky that those APIs weren't really used.

Added new tests.

HansMuller commented on packages/flutter/lib/src/widgets/routes.dart:
```diff
@@ -114,6 +114,12 @@ abstract class TransitionRoute<T> extends OverlayRoute<T> {
 
+  /// The animation for the route being pushed on top of this route. This
+  /// animation lets this route coordinate with the entrance and exit transition
+  /// of routes pushed on top of this route.
```
of the route being pushed on top of 

Doesn't the secondary animation also apply to the route that's being exposed because the route above it is being popped?

xster replied:
I just moved this block from below but all these docs are wrong. Re-editing them.

HansMuller replied:
It's not "routes pushed on top of this route" plural, it's just the one route pushed on top of this one?

xster replied:
Thanks, indeed it should be singular

HansMuller commented on packages/flutter/test/cupertino/route_test.dart:
```diff
@@ -309,23 +309,207 @@ void main() {
-    // and then press the button. MaterialPageTransition duration is 300ms,
-    // 275 = 300 * 0.75.
+    // Run the dismiss animation 60%, which exposes the route "push" button,
+    // and then press the button. A drag dropped animation is 400ms when dropped
```
What is a "drag dropped animation" and where does 400ms come from?

xster replied:
I moved this comment further down

HansMuller commented on packages/flutter/test/cupertino/route_test.dart:
```diff
@@ -309,23 +309,207 @@ void main() {
     await gesture.up();
-    await tester.pump(const Duration(milliseconds: 275)); // partially dismiss "route"
-    expect(find.text('route'), findsOneWidget);
+    await tester.pump(); // Trigger the dropped snapping animation.
```
What is a "dropped snapping" animation?

Generally speaking I think it would be helpful to explain this stuff in terms that would make sense to a developer who only knows that it's possible to drag a platform-iOS route from the left edge to initiate a navigator pop.

xster replied:
Done

HansMuller commented on packages/flutter/test/cupertino/route_test.dart:
```diff
@@ -309,23 +309,207 @@ void main() {
+    );
+
+    tester.state<NavigatorState>(find.byType(Navigator)).push(route2);
+
```
It would be helpful to explain that the numbers below are based on animating a Foo curve over a duration of Bar

xster replied:
Doh, found #29068 as I looked into it

xster replied:
Added more comments here

HansMuller commented on packages/flutter/test/cupertino/route_test.dart:
```diff
@@ -309,23 +309,207 @@ void main() {
+    expect(tester.getTopLeft(find.text('2')).dx, moreOrLessEquals(659, epsilon: 1));
+  });
+
+  testWidgets('Dragged pop gesture is linear', (WidgetTester tester) async {
```
Why are all of the epsilons needed if we're not curving the animation?

xster replied:
Just to avoid dealing with all the irrational numbers from dividing by 3. But it's only needed on the bottom route. Tweaking.

HansMuller reviewed (commented):
This looks mostly fine although I think future generations will appreciate a little less lingo.

xster committed:
review

goderbauer reviewed (commented):
This looks good to me. I'll leave the final review to Hans.

HansMuller commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -180,22 +180,19 @@ class CupertinoPageRoute<T> extends PageRoute<T> {
-  /// True if a Cupertino pop gesture is currently underway for [route].
+  /// True if an iOS-style back swipe pop gesture is currently underway for [route].
+  ///
+  /// This just check for the route's [NavigatorState.userGestureInProgress].
```
This just checks the route's ...

xster replied:
done

HansMuller commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -180,22 +180,19 @@ class CupertinoPageRoute<T> extends PageRoute<T> {
-  static bool isPopGestureInProgress(PageRoute<dynamic> route) => _popGestureInProgress.contains(route);
-  static final Set<PageRoute<dynamic>> _popGestureInProgress = <PageRoute<dynamic>>{};
+  static bool isPopGestureInProgress(PageRoute<dynamic> route) {
+    return route.navigator.userGestureInProgress;
```
I'm surprised that we need this method, but OK. I like the fact that it's possible to just ask the navigator.

xster replied:
We absolutely don't need this method since it adds more indirection confusion. But we can't break the API.

HansMuller reviewed (approved):
Sorry about the delay, LGTM

xster committed:
review
